docs/kittengrad.md

```diff
@@ -0,0 +1,151 @@
+# The Kittengrad Transactor Network
+
+WIP DRAFT
+
+## Introduction
+
+In this draft we discuss transactor network architecture and protocols for the
+next phase in the development of the Mediachain. The goal is to provide a baseline
+implemention that can scale to 100MM artefacts and beyond as we further develop and
+decentralize the system.
+
+The low level Mediachain architecture as it has evolved so far has two main components:
+the Transactor network, responsible for performing reads and writes to the
+mediachain, and the Datastore, responsible for storing mediachain data.
+Mediachain clients interact with the transactors through an rpc service interfaces.
+
+In the first phase, the transactor network was implemented as a Raft cluster
+using the copycat framework, providing a synchronous strongly linearized
+view of the mediachain. This was appropriate for prototyping purposes, but not
+applicable to the decentralized peer-to-peer architecture envisioned for the
+mediachain at scale.
+
+The Kittengrad architecture replaces the Copycat cluster with a peer-to-peer
+network of transactors, who independently accept and propagate transactions
+and synchronize with the generation of new blocks.
+This architecture no longer provides a synchronous view of the mediachain journal.
+Instead transactors initiate asynchronous transactions on behalf of clients,
+with eventual commit in the blockchain through a Block Commit Protocol.
+
+## Transactor Interface
+
+The transactor interface for clients is implemented through a standard rpc protocol
+(GRPC) with the following interface:
+
+```
+interface Transactor
+ insertCanonical(data): (TransactionId, CanonicalReference)
+ updateChain(CanonicalReference, data): TransactionId
+ lookupChain(CanonicalReference): ChainReference
+ currentBlockchain(): BlockReference
+ journalStream(): stream BlockReference
+```
+
+The read operations (`lookupChain`, `currentBlockchain`) are served locally by
+transactors. The write operations (`insertCanonical`, `updateChain`) initiate
+transactions that are completed asynchronously with confirmation streamed through
+the `journalStream` interface.
+
+When the transactor has built a backlog of transactions, it may
+reject an operation as temporarily unavailable. The client is expected to
+backoff and retry with a randomized exponential backoff or similar mechanism.
+
+The transactor will see through the completion of accepted transaction at a best
+effort basis, which in the absence of transactor failure should result in the eventual
+commit of an accepted transaction.
+The client may wait to confirm a transaction through a streamed block,
+and potentially retry transactions aborted because of transactor failure.
+
+The interface changes from the v0.1 interface are related to the asynchronous nature
+of write operations. Previously, the operations returned a commit receipt in the
```

Comment:
FWIW, the tx receipt for Tendermint looks like this: https://github.com/eris-ltd/eris-db/blob/develop-tmsp/rpc/core/types/responses.go#L102

Not entirely clear how you're thinking about this from the text, but I think the idea is that you get a tx id/hash back immediately upon acceptance, then listen for new blocks and see if your tx is in them as the confirmation mechanism.